oximeter/db/src/oxql/ast/grammar.rs

```diff
@@ -0,0 +1,1125 @@
+        ] {
+            let parsed = query_parser::query(q)
+                .unwrap_or_else(|_| panic!("failed to parse query: '{q}'"));
+            println!("{parsed:#?}");
+        }
+    }
+
+    #[test]
+    fn test_reducer() {
+        assert_eq!(query_parser::reducer("mean").unwrap(), Reducer::Mean);
+        assert!(query_parser::reducer("foo").is_err());
+    }
+
+    #[test]
+    fn test_parse_literal_timestamp_string() {
+        assert_eq!(
+            query_parser::timestamp_string("@2020-01-01").unwrap(),
+            Utc.with_ymd_and_hms(2020, 1, 1, 0, 0, 0).unwrap(),
+        );
+        assert_eq!(
+            query_parser::timestamp_string("@01:01:01").unwrap().time(),
+            NaiveTime::from_hms_opt(1, 1, 1).unwrap(),
+        );
+        assert_eq!(
+            query_parser::timestamp_string("@01:01:01.123456").unwrap().time(),
+            NaiveTime::from_hms_micro_opt(1, 1, 1, 123456).unwrap(),
+        );
+        assert_eq!(
+            query_parser::timestamp_string("@2020-01-01T01:01:01.123456")
+                .unwrap(),
+            NaiveDateTime::new(
+                NaiveDate::from_ymd_opt(2020, 1, 1).unwrap(),
+                NaiveTime::from_hms_micro_opt(1, 1, 1, 123456).unwrap(),
+            )
+            .and_utc(),
+        );
+    }
+
+    #[test]
+    fn test_parse_ipv4_literal() {
+        let check = |s: &str, addr: IpAddr| {
+            let Literal::IpAddr(ip) = query_parser::ip_literal(s).unwrap()
+            else {
+                panic!("expected '{}' to be parsed into {}", s, addr);
+            };
+            assert_eq!(ip, addr);
+        };
+        check("\"100.100.100.100\"", Ipv4Addr::new(100, 100, 100, 100).into());
+        check("\"1.2.3.4\"", Ipv4Addr::new(1, 2, 3, 4).into());
+        check("\"0.0.0.0\"", Ipv4Addr::UNSPECIFIED.into());
+
+        assert!(query_parser::ip_literal("\"abcd\"").is_err());
+        assert!(query_parser::ip_literal("\"1.1.1.\"").is_err());
+        assert!(query_parser::ip_literal("\"1.1.1.1.1.1\"").is_err());
+        assert!(query_parser::ip_literal("\"2555.1.1.1\"").is_err());
+        assert!(query_parser::ip_literal("1.2.3.4").is_err()); // no quotes
+    }
+
+    #[test]
+    fn test_parse_ipv6_literal() {
+        let check = |s: &str, addr: IpAddr| {
+            let Literal::IpAddr(ip) = query_parser::ip_literal(s).unwrap()
+            else {
+                panic!("expected '{}' to be parsed into {}", s, addr);
+            };
+            assert_eq!(ip, addr);
+        };
+
+        // IPv6 is nuts, let's just check a few common patterns.
+        check("\"::1\"", Ipv6Addr::LOCALHOST.into());
+        check("\"::\"", Ipv6Addr::UNSPECIFIED.into());
+        check("\"fd00::1\"", Ipv6Addr::new(0xfd00, 0, 0, 0, 0, 0, 0, 1).into());
+        check(
+            "\"fd00:1:2:3:4:5:6:7\"",
+            Ipv6Addr::new(0xfd00, 1, 2, 3, 4, 5, 6, 7).into(),
+        );
+
+        // Don't currently support IPv6-mapped IPv4 addresses
```

Comment:
A very good question. I need to revisit this, I think. We do currently store all addresses in ClickHouse as IPv6, storing IPv4 as mapped addresses. I'm not sure I know what happens if we run a filter that results in comparing a `std::net::Ipv4Addr` against that IPv6-mapped IPv4 address in the database!

Reply:
Ok, I think I've got this one done. Since we store all addresses as IPv6, it's actually pretty easy. When we run a filter in the DB, we just need to convert whatever IP is in the filter expression using ClickHouse's `toIPv6()` function. We were previously converting to V6 _or_ V4, based on the type, but that's wrong. They're always IPv6. We still just use the raw `std::net::IpAddr` when comparing in Rust, since we convert values from the DB to that automatically during deserialization.

This is done in 9ec70a23a.